### PR TITLE
Add daily and weekly to custom reports interval list

### PR DESCRIPTION
**Pull request by carkom**
~~Just a couple tweaks to get weekly working. Most the backend groundwork has already been laid.~~

~~Needs to "yearly" PR merged first.~~ Done.

Weekly became more complicated as I implented user prefs to match the begining of the week for the reports

### CODE REVIEW

**Comment by youngcw** [comment]
Could this pull in the setting for first day of the week and use that in the interval?

**Comment by Teprifer** [comment]
+1 to youngcw's comment above

Could the interval option be disabled when a graph is using the **Total** mode? It has no effect.

Likely more one for your future list, but a Fortnightly interval would be neat (lines up with a pay period).

_Just an aside, as entirely self afflicted, for mad people like myself, it takes a while, and Firefox asked if I wanted to wait or close the page, but it did eventually load daily for 7 years of data!_

**Comment by carkom** [comment]
> Could the interval option be disabled when a graph is using the **Total** mode? It has no effect.

Interval can still be used for many graphs in total mode. I think they are useable for many users. 

> Likely more one for your future list, but a Fortnightly interval would be neat (lines up with a pay period).

I'll have a look. 

> _Just an aside, as entirely self afflicted, for mad people like myself, it takes a while, and Firefox asked if I wanted to wait or close the page, but it did eventually load daily for 7 years of data!_

Yea, not recommended, but good to know it works!

File: packages/loot-core/src/shared/months.ts [hunk 1]
```diff
@@ -4,6 +4,26 @@ import memoizeOne from 'memoize-one';
 
 import * as Platform from '../client/platform';
 
+function isoWeekSettings(
+  date: Date,
+  type: string,
+  firstDayOfWeekIdx: number,
+): Date {
+  const checkNull = firstDayOfWeekIdx === undefined ? 1 : firstDayOfWeekIdx + 1;
+  const subNum =
```
Comment (carkom) [inline packages/loot-core/src/shared/months.ts:13]: This quite possibly the ugliest code I've ever written (it does, however, work as intended). I couldn't figure out any other way to get the default ISO Day to change off of Monday. If anyone has suggestions I'd be happy to use them.

**Review by youngcw — DISMISSED** [review]

**Review by twk3 — COMMENTED** [review]
@carkom can we reduce some of places we need to pass the firstDayOfWeekIdx around?

It looks like outside of loot core we could switch to importing it in `reportRanges.ts, grouped-spreadsheet.ts, custom-spreadsheets.ts` and reduce the need to keep passing it around in the higher level components.

File: packages/desktop-client/src/components/reports/ReportSidebar.jsx [hunk 2]
```diff
@@ -325,7 +329,7 @@ export function ReportSidebar({
               options={ReportOptions.dateRange
                 .filter(f => f[customReportItems.interval])
                 .map(option => [option.description, option.description])}
-              line={customReportItems.interval === 'Monthly' && dateRangeLine}
+              line={dateRangeLine > 0 && dateRangeLine}
```
[the remark this answers is not in the document]
Comment (carkom) [reply packages/desktop-client/src/components/reports/ReportSidebar.jsx:332]: This wouldn't work as it give a boolean result and the result that's needed is a number.

**Comment by carkom** [comment]
> @carkom can we reduce some of places we need to pass the firstDayOfWeekIdx around?
> 
> It looks like outside of loot core we could switch to importing it in `reportRanges.ts, grouped-spreadsheet.ts, custom-spreadsheets.ts` and reduce the need to keep passing it around in the higher level components.

Can't call it inside the spreadsheet since they are called with "useMemo" it comes back with a hooks error. 

As for reportRanges I get an "invalid Hook Call" if I try to use it in the function directly.

File: packages/desktop-client/src/components/reports/ReportSidebar.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (twk3) [reply packages/desktop-client/src/components/reports/ReportSidebar.jsx:332]: Got it, somehow missed that line wasn't just a flag

**Comment by twk3** [comment]
> Can't call it inside the spreadsheet since they are called with "useMemo" it comes back with a hooks error.
>
> As for reportRanges I get an "invalid Hook Call" if I try to use it in the function directly.

Ah, right. These aren't react components at this level, so we can't reference hooks in them. Apologies for the noise.

File: packages/loot-core/src/server/aql/compiler.ts [hunk 3]
```diff
@@ -600,6 +626,14 @@ const compileFunction = saveStack('function', (state, func) => {
     }
 
     // date functions
+    case '$day': {
+      validateArgLength(args, 1);
+      return castInput(state, args[0], 'date');
+    }
+    case '$week': {
```
Comment (twk3) [inline packages/loot-core/src/server/aql/compiler.ts:633]: I'm having trouble finding where this is used. I see us using `$day` in makeQuery, but not `$week`. Am I missing something?

File: packages/loot-core/src/shared/months.ts (hunk 1 above)
Comment (twk3) [reply packages/loot-core/src/shared/months.ts:13]: @carkom why do we need it to be the ISO week? Can't we just use `d.startOfWeek(date, {weekStartsOn: firstDayOfWeekIdx)` and `d.endOfWeek(date, {weekStartsOn: firstDayOfWeekIdx})` ?

Being that we are working with a Date object by this point, my hope is that the function will work as we intend?

File: packages/loot-core/src/server/aql/compiler.ts (hunk 3 above)
Comment (carkom) [reply packages/loot-core/src/server/aql/compiler.ts:633]: Good call out. I had it in there at first but removed the reference when I added the firstDayOfWeekIdx. Cheers!

File: packages/loot-core/src/server/aql/compiler.ts [hunk 4]
```diff
@@ -312,6 +321,23 @@ function castInput(state, expr, type) {
         'date-year',
       );
     }
+  } else if (type === 'date-week') {
```
Comment (twk3) [inline packages/loot-core/src/server/aql/compiler.ts:324]: Should we drop this as well if it's not being used?

Comment (carkom) [reply packages/loot-core/src/server/aql/compiler.ts:324]: Yep, shoulda noticed that. Thanks!

File: packages/loot-core/src/shared/months.ts (hunk 1 above)
Comment (carkom) [reply packages/loot-core/src/shared/months.ts:13]: Ahh! Yes! Thank you. I couldn't get out of my own head. Needed an outside set of eyes to diagnose it. Super elegant and easy solution. Thanks for the assist! 🧙‍♂️

**Review by twk3 — APPROVED** [review]
Has been working well in my testing. :+1:

**Pull request merged**